File: v3/src/core/distance_rings.py
"""
Precomputed distance rings for efficient move generation.
Critical for high-performance MCTS implementation.
"""

from typing import List, Tuple, Dict
import time

class DistanceRings:
    """
    Precomputed distance rings for 19x19 board.
    
    Instead of calculating distances during MCTS (O(n) per query),
    we precompute all distances once at startup (~50ms) and get O(1) lookups.
    
    This enables efficient progressive widening based on distance from existing stones.
    """
    
    def __init__(self, board_size: int = 19):
        self.board_size = board_size
        self.max_distance = board_size - 1  # Maximum possible distance on board
        
        # rings[position][distance] = list of positions at exactly that distance
        # Position is encoded as row * board_size + col for efficiency
        self.rings: Dict[int, Dict[int, List[Tuple[int, int]]]] = {}
        
        print(f"Precomputing distance rings for {board_size}x{board_size} board...")
        start_time = time.time()
        self._precompute_all_distances()
        end_time = time.time()
        print(f"Distance rings computed in {(end_time - start_time) * 1000:.1f}ms")
    
    def _precompute_all_distances(self):
        """Precompute distance rings for all positions."""
        for row in range(self.board_size):
            for col in range(self.board_size):
                position_id = self._encode_position(row, col)
                self.rings[position_id] = {}
                
                # Initialize distance buckets
                for distance in range(1, self.max_distance + 1):
                    self.rings[position_id][distance] = []
                
                # Calculate distances to all other positions
                for other_row in range(self.board_size):
                    for other_col in range(self.board_size):
                        if other_row == row and other_col == col:
                            continue  # Skip same position
                        
                        # Use Chebyshev distance (max of row/col differences)
                        # This is appropriate for Pente as stones can affect diagonally
                        distance = max(abs(other_row - row), abs(other_col - col))
                        
                        if distance <= self.max_distance:
                            self.rings[position_id][distance].append((other_row, other_col))
    
    def _encode_position(self, row: int, col: int) -> int:
        """Encode position as single integer for efficient storage."""
        return row * self.board_size + col
    
    def get_positions_at_distance(self, center_row: int, center_col: int, distance: int) -> List[Tuple[int, int]]:
        """
        Get all positions at exactly the specified distance from center.
        
        Args:
            center_row: Center row position
            center_col: Center col position  
            distance: Exact distance to query
            
        Returns:
            List of (row, col) positions at that distance
        """
        if not (0 <= center_row < self.board_size and 0 <= center_col < self.board_size):
            return []
        
        if distance < 1 or distance > self.max_distance:
            return []
        
        position_id = self._encode_position(center_row, center_col)
        return self.rings[position_id][distance]
    
    def get_positions_within_distance(self, center_row: int, center_col: int, max_distance: int) -> List[Tuple[int, int]]:
        """
        Get all positions within the specified maximum distance from center.
        
        Args:
            center_row: Center row position
            center_col: Center col position
            max_distance: Maximum distance (inclusive)
            
        Returns:
            List of (row, col) positions within that distance, ordered by distance
        """
        if not (0 <= center_row < self.board_size and 0 <= center_col < self.board_size):
            return []
        
        if max_distance < 1:
            return []
        
        positions = []
        position_id = self._encode_position(center_row, center_col)
        
        for distance in range(1, min(max_distance + 1, self.max_distance + 1)):
            positions.extend(self.rings[position_id][distance])
        
        return positions
    
    def get_distance(self, row1: int, col1: int, row2: int, col2: int) -> int:
        """
        Get Chebyshev distance between two positions.
        
        This is a simple calculation but provided for completeness.
        For performance-critical code, consider caching if needed.
        """
        return max(abs(row2 - row1), abs(col2 - col1))
    
    def get_ordered_positions_around_stones(self, stone_positions: List[Tuple[int, int]], 
                                           max_distance: int = 3) -> List[Tuple[int, int]]:
        """
        Get positions ordered by minimum distance to any existing stone.
        
        This is the key method for efficient move generation in MCTS:
        - Positions closer to existing stones are returned first
        - Duplicates are removed 
        - Critical for progressive widening
        
        Args:
            stone_positions: List of (row, col) positions with existing stones
            max_distance: Maximum distance to consider
            
        Returns:
            List of positions ordered by distance to nearest stone
        """
        if not stone_positions:
            return []
        
        # Track positions by their minimum distance to any stone
        distance_buckets: Dict[int, List[Tuple[int, int]]] = {}
        seen_positions = set()
        stone_positions_set = set(stone_positions)  # For quick lookup
        
        # Initialize distance buckets
        for distance in range(1, max_distance + 1):
            distance_buckets[distance] = []
        
        # For each existing stone, add positions at each distance
        for stone_row, stone_col in stone_positions:
            if not (0 <= stone_row < self.board_size and 0 <= stone_col < self.board_size):
                continue
                
            position_id = self._encode_position(stone_row, stone_col)
            
            for distance in range(1, max_distance + 1):
                if distance > self.max_distance:
                    break
                    
                for pos in self.rings[position_id][distance]:
                    if pos not in seen_positions and pos not in stone_positions_set:
                        # This position hasn't been seen at a closer distance and isn't a stone
                        distance_buckets[distance].append(pos)
                        seen_positions.add(pos)
        
        # Combine buckets in distance order
        ordered_positions = []
        for distance in range(1, max_distance + 1):
            ordered_positions.extend(distance_buckets[distance])
        
        return ordered_positions
    
    def get_statistics(self) -> Dict[str, int]:
        """Get statistics about the precomputed rings."""
        total_positions = self.board_size * self.board_size
        total_ring_entries = 0
        
        for position_id in self.rings:
            for distance in self.rings[position_id]:
                total_ring_entries += len(self.rings[position_id][distance])
        
        return {
            'board_size': self.board_size,
            'total_positions': total_positions,
            'max_distance': self.max_distance,
            'total_ring_entries': total_ring_entries,
            'memory_positions': len(self.rings)
        }
    
    def __str__(self) -> str:
        """String representation with statistics."""
        stats = self.get_statistics()
        return (f"DistanceRings({stats['board_size']}x{stats['board_size']}, "
                f"{stats['total_ring_entries']} precomputed distances)")

# Global instance for efficiency - created once at startup
_global_distance_rings: DistanceRings = None

def get_distance_rings(board_size: int = 19) -> DistanceRings:
    """
    Get global distance rings instance.
    Creates it once at startup for efficiency.
    """
    global _global_distance_rings
    if _global_distance_rings is None or _global_distance_rings.board_size != board_size:
        _global_distance_rings = DistanceRings(board_size)
    return _global_distance_rings

File: src/core/distance_rings.cpp
#include "core/distance_rings.hpp"
#include <unordered_set>
#include <algorithm>

namespace core {

DistanceRings::DistanceRings() {
    precompute_all_distances();
}

void DistanceRings::precompute_all_distances() {
    // For each position on the board
    for (int center_row = 0; center_row < BOARD_SIZE; ++center_row) {
        for (int center_col = 0; center_col < BOARD_SIZE; ++center_col) {
            Position center{static_cast<int8_t>(center_row), static_cast<int8_t>(center_col)};
            int center_id = position_to_id(center);
            
            // For each other position on the board
            for (int target_row = 0; target_row < BOARD_SIZE; ++target_row) {
                for (int target_col = 0; target_col < BOARD_SIZE; ++target_col) {
                    Position target{static_cast<int8_t>(target_row), static_cast<int8_t>(target_col)};
                    
                    // Skip the center position itself
                    if (center == target) continue;
                    
                    int distance = chebyshev_distance(center, target);
                    
                    // Only store distances within our maximum range
                    if (distance <= MAX_DISTANCE) {
                        rings_[center_id][distance].push_back(target);
                    }
                }
            }
            
            // Sort each distance ring for consistent ordering
            for (int distance = 1; distance <= MAX_DISTANCE; ++distance) {
                std::sort(rings_[center_id][distance].begin(), rings_[center_id][distance].end(),
                    [](const Position& a, const Position& b) {
                        if (a.row != b.row) return a.row < b.row;
                        return a.col < b.col;
                    });
            }
        }
    }
}

const std::vector<Position>& DistanceRings::get_positions_at_distance(
    Position center, int distance) const noexcept {
    
    // Bounds checking
    if (center.row < 0 || center.row >= BOARD_SIZE || 
        center.col < 0 || center.col >= BOARD_SIZE ||
        distance < 0 || distance > MAX_DISTANCE) {
        static const std::vector<Position> empty;
        return empty;
    }
    
    int center_id = position_to_id(center);
    return rings_[center_id][distance];
}

std::vector<Position> DistanceRings::get_ordered_moves_around_stones(
    const std::vector<Position>& stone_positions,
    int max_distance) const {
    
    std::unordered_set<int> unique_positions; // Use position_id for uniqueness
    std::vector<std::vector<Position>> distance_groups(max_distance + 1);
    
    // Add stone positions to unique_positions to exclude them from results
    for (const Position& stone : stone_positions) {
        unique_positions.insert(position_to_id(stone));
    }
    
    // Collect all positions at each distance from any stone
    for (const Position& stone : stone_positions) {
        for (int distance = 1; distance <= max_distance; ++distance) {
            const auto& positions_at_distance = get_positions_at_distance(stone, distance);
            
            for (const Position& pos : positions_at_distance) {
                int pos_id = position_to_id(pos);
                
                // Only add if we haven't seen this position before
                if (unique_positions.find(pos_id) == unique_positions.end()) {
                    unique_positions.insert(pos_id);
                    distance_groups[distance].push_back(pos);
                }
            }
        }
    }
    
    // Build result vector ordered by distance (closest first)
    std::vector<Position> result;
    for (int distance = 1; distance <= max_distance; ++distance) {
        // Sort positions at this distance for consistency
        std::sort(distance_groups[distance].begin(), distance_groups[distance].end(),
            [](const Position& a, const Position& b) {
                if (a.row != b.row) return a.row < b.row;
                return a.col < b.col;
            });
        
        // Add to result
        result.insert(result.end(), distance_groups[distance].begin(), distance_groups[distance].end());
    }
    
    return result;
}

} // namespace core